python/mxnet/gluon/probability/distributions/distribution.py

```diff
@@ -0,0 +1,196 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# coding: utf-8
+# pylint: disable=wildcard-import
+"""Base distribution class."""
+__all__ = ['Distribution']
+
+from .utils import cached_property
+from numbers import Number
+
+
+class Distribution(object):
+    r"""Base class for distribution.
+    
+    Parameters
+    ----------
+    F : mx.ndarray or mx.symbol.numpy._Symbol
+        Variable that stores the running mode.
+    event_dim : int, default None
+        Variable indicating the dimension of the distribution's support.
+    validate_args : bool, default None
+        Whether to validate the distribution parameters
+    """          
+
+    # Variable indicating whether the sampling method has
+    # pathwise gradient.
+    has_grad = False
+    support = None
+    has_enumerate_support = False
+    arg_constraints = {}
+    _validate_args = False
+
+    @staticmethod
+    def set_default_validate_args(value):
+        if value not in [True, False]:
+            raise ValueError
+        Distribution._validate_args = value
+
+    def __init__(self, F=None, event_dim=None, validate_args=None):
+        self.F = F
```

Comment:
Should the F be removed as it's no longer needed?

Reply:
Why is it no longer needed? Do you mean it will be replaced by deferred computation?

Reply:
Yes. Because eventually we like to refactor the backend and replace the nnvm representation. The first step is to remove the explicit reference from the frontend, which implies removing F.

Reply:
In the original design back to last year, we take F into account, and F is not used in every part of the module. If we need to remove F, it will take a lot of effort to redesign. I think removing F can be left to the future.

Reply:
> In the original design back to last year, we take F into account, and F is used in every part of the module.

Could you post the reference? Ideally this would have been caught early in a design proposal document. I believe I discussed the plans with Xi in person before.

> If we need to remove F, it will take a lot of effort to redesign.

This PR currently targets the master branch and there wouldn't be any release with the F if we follow https://github.com/apache/incubator-mxnet/issues/17676. So this effort may need to be spent in any case. Now or in the future. However, I think the effort required to fix this PR wouldn't be "a lot". But if it's currently not possible, would it be better to target the 1.x branch and port to the master branch once resources are available?

Reply:
When we started doing this, it was November last year, and Xi told me you mentioned that it is not guaranteed that F can be removed before the release of 2.0. So we did everything based on the original hybrid block, and we cannot test our design if we followed the new api because it was not in MXNET at that time. Also, when the deferred compute was merged, we basically finished all the module except some functionalities. I remembered that I talked to @szha that we can keep F in probability module before.

Reply:
@leezu 
The distribution module kicked off last year, and we had most of the features finished around early January this year when our RFC was opened as one of the items for mxnet 2.0: https://github.com/apache/incubator-mxnet/issues/17240. If we take a look at the RFC for deferred compute, we can see that this RFC was still getting concerns and comments from the community members by late January: https://github.com/apache/incubator-mxnet/issues/16376#issuecomment, and the RFC for deprecating `F` was out late February: https://github.com/apache/incubator-mxnet/issues/17676.
On the other hand, we did respect your effort on deferred compute when we started designing this module, and I remember during one of our earlier offline exchanges, you were not confident enough about finishing this deferred compute feature before 2.0 release. With that piece of information, we eventually decided to proceed with the traditional Gluon (with `F`) in this module. Also, just like what @szhengac said, he reached a consensus with @szha offline about keeping the `F` in this module.
I don’t think it’s reasonable to ask people to predict the future and onboard a new feature that was not guaranteed to be ready.

Reply:
@xidulu the timeline of implementation does not have anything to do with ensuring that the final PR integrates well into the overall project. Have you actually taken a look at what steps I requested you to make? I don't think so. The aim of my code-review request here is to make sure that the code you PR is integrated correctly with the Gluon API. For this, you'll need to make a few high-level changes in the way the StochasticBlock is called. Specifically,

- delete self.F = F
- use def forward instead of def hybrid_foward
- Change the checks of isinstance(Symbol) to checking if the Block is hybridized.

These are mostly trivial changes, but I agree there will be an effort to check if any unexpected problems occur. This is nothing unexpected in the case of parallel feature development. The recommendation to start developing against the hybrid_forward API is based on the fact that both APIs are highly related and thus the effort of the final migration is low.

From my perspective it's also fine to keep the F and target the 1.x branch in this PR. Then we may have a separate PR integrating into the master branch. Let me know if you prefer that and I'll be happy to do another review about any issues that the current PR may have with respect to the 1.x API.

Reply:
@leezu 

`StochasticBlock`, in essence, is just a block capable of accumulating loss during the forward propagation. I am very happy to make it more compatible with the new Gluon api. 

Secondly, I don’t think it’s necessary to remove `self.F` in distribution object. In our design, users **do not** need to manually specify `F` when creating a distribution object, as the running mode will be automatically determined by the `getF` function, for example 
https://github.com/apache/incubator-mxnet/pull/18403/files#diff-613421c6ce6fc8ec3338aa29c9535087R48. 

I also made a little modification to the test cases to demonstrate how it works in practice (as well as its compatibility with the new hybridBlock api) https://github.com/apache/incubator-mxnet/pull/18403/files#diff-27fee32587466a570c77da139b7136f6R94-R96

Reply:
@xidulu 

> To summarize, the F in the distribution object does not violate the 2.x API. Instead, it can fit well into both 1.x and 2.x API.

Note that the Gluon.HybridBlock in the master branch currently implements both forward and hybrid_forward for transition purpose. The hybrid_forward is deprecated and will be deleted from all HybridBlocks prior to the release. There was just too many code relying on the deprecated interface to delete it outright. This doesn't mean we should add more code relying on the deprecated interface.

Specifically, the `getF` function used in the StochasticBlock would never see a symbolic input. Thus part of the code in your PR can be simplified / updated. Sorry if this background to my previous comment wasn't completely clear.

Reply:
@leezu 

Maybe I did not make it clear in the previous comments, `Distribution` and `getF` both have nothing to do with `Block`, `HybridBlock` or `StochasticBlock`. A `Distribution` object can be used inside any types of blocks or even outside a block. The `getF` function is invoked in the initialization phase of a `Distribution` object to determine *whether to use `sym.np` or `nd.np` for the computation in the member function*, for example:
https://github.com/apache/incubator-mxnet/pull/18403/files#diff-e83cafb33610cdde004c72426d6d1492R82

In short, if the parameters are Symbols, then all the methods in the object will use function under the `sym.np` namespace. If the parameters are NDArray, then we'll choose to use functions under `nd.np`. 

I understand the possibility of `hybrid_forward` getting deleted in the future, but that deprecation hasn't reach community consensus, right? (https://github.com/apache/incubator-mxnet/issues/17676).

Furthermore, as I demonstrate in this test case: https://github.com/apache/incubator-mxnet/pull/18403/files#diff-27fee32587466a570c77da139b7136f6R94-R96, current implementation works perfectly well with the new Gluon API (notice that `F` is not specified in the initialization function), that is to say, distribution object itself **does not rely on** the deprecated interface at all.

What I can do at this moment is:
- Create a new test file based on the new Gluon API. 
- Raise warning to discourage users from explicitly passing `F` when creating a distribution object.

Once `hybrid_forward` got officially deleted from the code base, I would then consider replacing all the `F` in this module with `mx.np` as well as removing `F` from the initialization function.

------------------------------
Your suggestion on StochasticBlock is correct, current implementation is based on the old version of `HybridBlock`, it is necessary to make some modifications.

Reply:
@xidulu the consensus in the ticket you reference is to delete `mx.sym` and `mx.nd`. Only `mx.np` and `mx.npx` remain. That's why you don't need the F.